024-file-tokens.md

```diff
@@ -0,0 +1,199 @@
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail, keeping in mind the following:
+
+- Its interaction with other parts of the system is clear
+- It is reasonably clear how the contribution would be implemented
+- Dependencies on libraries, tools, projects, or work that isn't yet complete
+- New API routes that need to be created or modifications to the existing routes (if needed)
+- Any breaking changes and ways in which we can ensure backward compatibility.
+- Use Cases
+- Goals
+- Deliverables
+- Changes to documentation
+- Ways to scale the solution
+
+Ensure that you include examples and code snippets to allow the community to understand the proposed solution. **It would be best if the examples use naming conventions that you intend to use during the actual implementation to suggest changes early on during the development.**
+
+Write your answer below.
+
+-->
+
+### API Endpoints
+
+<!--
+List the new API routes or endpoints that we might need to add for supporting the new feature.
+Keep in mind to stay very strict to the API protocol and method, whether your new
+changes are for the REST, WebSocket or any other API protocol Appwrite supports.
+
+For example:
+
+**POST /v1/coffee ** - an endpoint for creating coffee.
+**DELETE /v1/coffee ** - an endpoint for deleting coffee.
+-->
+
+Tokens will be a part of storage service
+
+**POST /v1/storage/buckets/:bucketId/files/:fileId/tokens** -           create token
+**PUT /v1/storage/buckets/:bucketId/files/:fileId/tokens** -            update token
+**GET /v1/storage/buckets/:bucketId/files/:fileId/tokens** -            List token
+**GET /v1/storage/buckets/:bucketId/files/:fileId/tokens/:tokenId** -   Get token
+**DELETE /v1/storage/buckets/:bucketId/files/:fileId/tokens/:tokenId** -         delete token
+
+
+### Data Structure
+
+<!--
+What kind of changes or additions are required for the Appwrite base collections
+to support this feature. Explain which entities should be added or updated, what new attributes they
+need to have and why. Please think well about the naming conventions and how well they play with other
+Appwrite conventions. Try and stay as consistent with existing patterns as much as possible.
+-->
+
+We need to introduce new collection to save tokens.
+
+**tokens**
+  - bucketId - String
+  - fileId - String
+  - secret - String
+  - permissions - array string
+  - expiryDate - date time
+
+We will also need a token validator, that will validate that the token is not expired and has access to the resources.
+
+### Supporting Libraries
+
+<!--
+Which different libraries do we need to support the new features?
+Please describe the new library's potential API?
+Avoid using 3rd party libraries when possible, if required - explain why.
+-->
+
+This particular feature doesn't require any supporting libraries.
```

Comment:
We might need to use JWT tokens for validating the tokens and their permissions. This is because we might not be able to access our DB for validation when serving cached images.

Reply:
Not sure I understand completely